### ISSUE TITLE
[web] The future of `flutter_service_worker.js`

### ISSUE THREAD

**Issue by ditman** [flutter/flutter#156910]
### Document Link

https://flutter.dev/go/web-cleanup-service-worker

### What problem are you solving?

The current default service worker implementation is one of (several) approaches that can be used by Flutter web, however it's not a good fit for all use-cases.

This confuses users when it doesn't behave as they expect.

Having a default service worker implementation sends the message that it is *necessary* or *recommended* by Flutter web to function correctly, but that is not the case.

This design doc discusses how the current implementation of `flutter_service_worker.js` is going to be phased out (TL;DR: with a cleanup service worker for those currently using the default one), and how the flutter tools and flutter.js will eventually stop generating (or loading) a default service worker, allowing web app developers to do what's best for their specific apps.

**Comment by slavap** [comment]
@ditman 
>> and flutter.js will eventually stop generating (or loading) a default service worker, allowing web app developers to do what's best for their specific apps.

the best for any web app is to have PWA "just working". If someone for some reason wants "custom PWA", let them do it, but for others 90% of "lazy" developers - the default implementation of "one of (several) approaches" is absolutely fine, just make it "dumb" and stable as AK-47.

**Comment by uldall** [comment]
@slavap I do not believe a PWA requires a service worker now a days. See [https://web.dev/articles/install-criteria#criteria](https://web.dev/articles/install-criteria#criteria).

**Comment by slavap** [comment]
@uldall if that is correct and PWA could work without service worker - then that has to be clearly explained in the document.

**Comment by ditman** [comment]
@slavap thanks for your message, feedback is always appreciated! (you should also be able to comment on the google doc itself if you want to!)

To me, the biggest problem that the current default service worker has is that **people think that it is _mandatory_**, when in reality, it isn't.

For most of Flutter web developers, the standard browser cache will work as expected, and they'll have a better development and deployment experience. (The rest 1% is providing their own service worker implementation anyway to achieve whatever extra goals they have.)

I'm not opposed to generating service workers with Flutter web, but I think we should:

* have a very clear idea of *what the service worker is supposed to do*: What happens when it goes offline? How does it react to updates? How does it cache assets? etc... I don't think we have any of this right now.
* make it an opt-in feature (maybe make users manually register the `flutter_service_worker.js`, instead of doing it automatically)
* (Ideally) be able to use battle-tested technology that already exists (like [Workbox](https://developer.chrome.com/docs/workbox) or any other), and for that...
* (Ideally) have the `flutter` tooling play _nicer_ with other build systems

I don't know how to get to that future yet! Maybe it can be done as a [package:build](https://pub.dev/packages/build), or using the upcoming [Native assets feature](https://github.com/dart-lang/sdk/issues/50565), but I think it's better to cleanup and then try again.

In the meantime, users who *really* need a service worker can build their own, without worrying that Flutter is going to step on their toes.

---

PS: There's cool reads about "PWA", service workers and related stuff here:
* https://web.dev/explore/progressive-web-apps

**Comment by slavap** [comment]
@ditman 
>> (Ideally) be able to use battle-tested technology that already exists (like [Workbox](https://developer.chrome.com/docs/workbox) or any other), and for that...

This sounds like a totally right idea to me, integrating with "battle-tested technology" and not reinventing the wheel is always great.

**Comment by jezell** [comment]
Love this change!

**Comment by lukepighetti** [comment]
I do think it's reasonable for users to expect Flutter to provide a sane default PWA experience (like every other popular web framework). 

But IIRC that is blocked right now by issues like hotfixes not working as expected.

So any short term feature losses that resolve the big issues deploying flutter web are welcome... as long as the vision for a sane default PWA experience is retained.

**Comment by yjbanov** [comment]
@ditman Can you please clarify in the doc what the plan is for the `manifest.json` file?

**Comment by rayliverified** [comment]
Does this remove cache busting? I have a concern that this will make updating Flutter Web apps successfully even harder.
One issue that's popped up with Flutter Wasm builds is that something gets cached by the browser and stuck so newer builds often fail. New builds try to reference some symbols that are different from old builds.

Refreshing the page does not fix. Clearing cache doesn't fix either. The only fix is "Clear application data" in dev tools.
![image](https://github.com/user-attachments/assets/d51f896b-caff-471f-9071-f0f50b8c8d92)

How does this impact the updating web apps workflow?

One thing I can see it helping is it would let me run my own service worker to clear the application data when I detect a newer version of my app from the server. But that wouldn't be out of the box behavior. So what would the default behavior and strategy be?

Ideally, Flutter Web developers can publish a new version, the new version runs and works.
That's not what is happening now. Would that be a goal of this change?

**Comment by ditman** [comment]
> So what would the default behavior and strategy be?

@rayliverified The current implementation attempts to aggressively cache and pre-fetch the `assets` of your app; that will stop happening. The new default behavior will be "remove the default service worker, if it's installed.", so we can help "clean up" web apps that don't care about the service worker and just take "the default". Later, we will stop emitting a service worker completely so you can install your own.

> How does this impact the updating web apps workflow? [...] 
> Ideally, Flutter Web developers can publish a new version, the new version runs and works. That's not what is happening now. Would that be a goal of this change?

Yes, this change should move your app to using the default browser cache (that you can control via your CDN configuration), rather than having our own implementation on top of that. This should make your update-after-deployment experience much more straightforward, and was specifically one of the things that motivated this change!

**Comment by ditman** [comment]
> So any short term feature losses that resolve the big issues deploying flutter web are welcome... as long as the vision for a sane default PWA experience is retained.

@lukepighetti We definitely want to make Flutter web The Best Choice™ for PWAs (or at least be very good at them!)

**Comment by rayliverified** [comment]
Thanks for explaining. 

Giving the developer more control is always great.


Can you confirm the following:
In the updated scenario, the developer is responsible for updating the assets cache. This means if the developer updates an asset without changing the name, they will need to either implement a cache strategy or configure the CDN settings to the correct settings.


Image A (dolphin)

Developer replaces with Image A (hippo). Name remains the same.


CDN server sets Image A to Cache-Control: max-age=86400
App will serve old Image A (dolphin) to users until the cache expires.



The previous behavior is that the manifest.json will detect when the image change and make sure the updated one is loaded by appending a cache key.


Thank you for the additional clarity.

**Comment by ditman** [comment]
> App will serve old Image A (dolphin) to users until the cache expires.

This is not entirely true. For those assets that get reuploaded (everything in build/web pretty much), you should also use an `ETag` header (if you want to make this fully CDN driven).

This looks like a good read about this topic:

* https://simonhearne.com/2022/caching-header-best-practices/#the-solution

> The previous behavior is that the manifest.json will detect when the image change and make sure the updated one is loaded by appending a cache key.

The previous behavior is that the `flutter_service_worker.js` would contain a list of the assets + their content hash, so any changes to the contents of the assets would generate a slightly different service worker file, which would trigger an "update" flow of the service worker. In the "activate" event of the service worker, it'd compare the hash of every cached file with the new one, and if different, it'd [purge them from the cache](https://github.com/flutter/flutter/blob/53a97dd9d434476862ebaf864af7730f7e599bd8/packages/flutter_tools/lib/src/web/file_generators/js/flutter_service_worker.js#L53-L58).

(In fact, I suspect that one of the problems of the current service worker is that we're **always changing its name by appending a `?v=` version string to it**, which is [discouraged](https://web.dev/articles/service-worker-lifecycle#avoid-url-change))

**Comment by jroitgrund** [comment]
>Yes, this change should move your app to using the default browser cache (that you can control via your CDN configuration)

Can you be more specific about the intended usage without the service worker?

* Using `max-age` will allow users to see stale versions of the app, which is one of the main problems with the service worker.
* Using `must-revalidate` + `if-modified-since` / `if-none-match`:
  * always requires a round-trip to the server, which means the cache helps with bandwidth, but not with latency
  * isn't supported on some CDNs (e.g. firebase)

In the JS world, the standard for cache busting is for the HTML to reference `<script src=foo.js?version=<hash>`. Whenever the shipped assets change, the query parameter changes, which busts the cache. This means you can safely use `public,max-age` (avoiding server roundtrips) and yet be sure that users will always get the latest version of your assets.

**Comment by jroitgrund** [comment]
>Cache busting is related, but independent of this discussion

Currently, the service worker provides caching to prevent server roundtrips on subsequent loads. It feels important to discuss how users will be able to emulate this caching behavior and keep avoiding these server roundtrips without relying on the service worker. The current proposal doesn't address this.

>You can bust the cache before we append a cache-busting parameter by adding a version string to your asset names: logo.png -> logo.v6.png.

This only works for assets entirely controlled by developers, it doesn't work for all the flutter-generated JS, since all that bootstrapping code is generated by Flutter.

>I'm currently deploying stuff to Firebase Hosting, and it uses ETag headers to control the cache in my browser. Every time I redeploy and refresh, I get the new stuff, without the need for any busting (but YMMV?).

What Firebase *doesn't do* is ever return 304 if the headers *DO* match. Try it for yourself!

https://github.com/jroitgrund/does-firebase-cache / https://does-firebase-cache.web.app/ is a sample project, where I did:

* firebase init
* touch public/example.js
* firebase deploy

With curl:

```
$ curl -s -I -X GET 'https://does-firebase-cache.web.app/example.js' | egrep "(HTTP|etag)"
HTTP/2 200
etag: "f1b651238a58fe290baec6c5e32f3bdb1943dd2bd582f02569231f7a757c7837"
$ curl -s -I -X GET 'https://does-firebase-cache.web.app/example.js' -H 'If-None-Match: "f1b651238a58fe290baec6c5e32f3bdb1943dd2bd582f02569231f7a757c7837"' | egrep "(HTTP|etag)"
HTTP/2 200
etag: "f1b651238a58fe290baec6c5e32f3bdb1943dd2bd582f02569231f7a757c7837"
```

Contrast this with the expected behavior on a random site which does implement this correctly

```
$ curl -s -I -X GET 'https://app.teamsurfboard.com/' | egrep "(HTTP|Etag)"
HTTP/1.1 200 OK
Etag: W/"333-1930737cdf0"
$ curl -s -I -X GET 'https://app.teamsurfboard.com/' -H 'If-None-Match: W/"333-1930737cdf0"' | egrep "(HTTP|Etag)"
HTTP/1.1 304 Not Modified
Etag: W/"333-1930737cdf0"
```


If you were to add 

```
    "headers": [
      {
        "source": "**/*",
        "headers": [
          {
            "key": "Cache-Control",
            "value": "no-cache"
          }
        ]
      }
    ]
```

to the config (which is the desired setup to avoid users seeing stale code), users would *never* see a 304.

I realize that this is a Firebase issue and a not a Flutter issue, but it's hard to talk about removing the service worker in a vacuum, when the consequence will be:

* it's impossible to fully cache flutter-generated JS (fully as in no server roundtrip)
* it's impossible to cache flutter-generated JS AT ALL if you're hosting on Firebase (except using max-age headers which present the exact same problem as the service worker, and make it hard to guarantee all your users see an updated app when you see a new version).

**Comment by jroitgrund** [comment]
Agreed - these work, but they depend on post-processing the black box output of `flutter build web`, which isn't ideal.

**Comment by slavap** [comment]
@jroitgrund 
>> they depend on post-processing
Second workaround does not post-process anything.

**Comment by ditman** [comment]
> Note that the current waterfall (index.html -> flutter_bootstrap.js -> main.js) requires 3 round-trips!

@jroitgrund You can inline your `flutter_bootstrap.js` into your `index.html`, and save one of those round trips. See [the `{{flutter_bootstrap_js}}` token docs](https://docs.flutter.dev/platform-integration/web/initialization#customize-initialization).

**Comment by DavidOrakpo** [comment]
Is there any updates on this?

**Comment by kevmoo** [comment]
@DavidOrakpo – see the in-flight PR: https://github.com/flutter/flutter/pull/176834

**Comment by linghengqian** [comment]
Out of curiosity, since we've abandoned service workers for so long, couldn't we just create a PR to remove all the code related to `flutter_service_worker.js`? Or is there something I've overlooked? 

The documentation for https://flutter.dev/go/web-cleanup-service-worker was last updated in 2024.

**Comment by BirjuVachhani** [comment]
I am using `--pwa-strategy=none` flag to disable the PWA service worker to avoid caching issues. However I am seeing this warning that this flag is deprecated now. What's the alternative to disable this now?

```
The --pwa-strategy option is deprecated and will be removed in a future Flutter release.
For more information, see: https://github.com/flutter/flutter/issues/156910
```

**Comment by ditman** [comment]
@linghengqian there has been some work towards this, here: https://github.com/flutter/flutter/pull/176834. The removal is slow because of Flutter's [Deprecation Policy](https://docs.flutter.dev/release/compatibility-policy#deprecation-policy), it takes a while for everybody to catch up.

Anyway, I'm not on the web team anymore, @mdebbar and the team are driving this!

**Comment by ditman** [comment]
> However I am seeing this warning that this flag is deprecated now. What's the alternative to disable this now?

@BirjuVachhani the warning was to draw your attention to this issue. For now, it'll continue working as before. Once the whole machinery related to the service worker is completely removed, the command line option is going to stop working and you'll be able to stop using it, because flutter won't attempt to install any service worker at all.
